### DISCUSSION TITLE
Best option right now for embedded plotting?

### DISCUSSION THREAD

**Discussion by laserman781** [pyvista/pyvista#4778] [Q&A]
I am currently utilizing voila-vuetify in order to create a web app with PyVista. I would need to be able to have embedded plotting so that I could eventually move away from a desktop and do so on a tablet. What would be the best option to do this? Maybe exporting as vtkjs or vtp and using another package to plot within the notebook? 

Ideally, I'd want to have all the potions I have in the PyVista plotter where I can include the xyz axis, scalar bars, etc.

Thank you!

**Comment by banesullivan** [comment]
There isn’t really a “best option” when considering embedded renderings at this time - they all come with tradeoffs.

If your priority is having axes grids and scalar bars, then K3D might be the best fit at this time. Unfortunately, my experience with K3D is that it is pretty hacky and limited on features / the types of data in can display (well). I’d recommend checking out the todo list for items in this pull request (https://github.com/pyvista/pyvista/pull/300) and helping where you can with that PR to export PyVista scenes to K3D scenes. There’s definitely no elegant solution for what you want to do at this time but I think K3D holds a lot of potential (@marcomusy has worked to export `vtkplotter` scenes to K3D and might have some code we could borrow for that pull request)

Exporting as VTKjs using either the export script we have in PyVista or the one from `panel` will not have scalar bars, axes grids, labels, or anything other than the data itself.

To move forward, you’d definitely need https://github.com/pyvista/pyvista/pull/300 to land... unfortunately, I do not currently have the bandwidth to wrap it up.

Moving forward, I will be focusing my attention on building interoperability between `itkwidgets` and PyVista as it is a much better framework for web rendering in my opinion and holds potential for adding things like scalar bars, axes labels, and other widgets down the road.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by laserman781** [comment]
I haven't been able to get any of the solutions to work properly, is there maybe a way I can utilize simply show the 3d render for now? Whenever I do `p.show(auto_close=False,use_panel=True)` I get a static image. until the is better interoperability between `itkwidgets`  and Pyvista how can accomplish this? I've copy-pasted your panel sphere example and I only get a static image where in your example you get the embedded widget...

**Comment by banesullivan** [comment]
Do you have [`panel`](https://panel.pyviz.org) installed?

**Comment by marcomusy** [comment]
> There’s definitely no elegant solution for what you want to do at this time but I think K3D holds a lot of potential (@marcomusy has worked to export `vtkplotter` scenes to K3D and might have some code we could borrow for that pull request)

Hi, sure, you can pick from here if it's useful to pyvista:
https://github.com/marcomusy/vtkplotter/blob/23a0ff9a34e3812f81cc29248e41934bd251b2ea/vtkplotter/backends.py#L77

**Comment by banesullivan** [comment]
Thanks, @marcomusy! I got started making a PyVista to K3D exporter modeled after your work over in https://github.com/pyvista/pyvista/pull/300, but I quickly ran into issues around meshes with line segments (see https://github.com/K3D-tools/K3D-jupyter/issues/173) and other TODO items in https://github.com/pyvista/pyvista/pull/300

Hopefully, I'll have time to circle back to that PR before long!
